### PR TITLE
add better route parsing

### PR DESCRIPTION
**Pull request by trvswgnr**
`¯\_(ツ)_/¯`

### CODE REVIEW

**Comment by cod1r** [comment]
This PR is too good.

**Comment by trvswgnr** [comment]
@twny fixed the conflicts! i also moved the hashmap into the main function so it doesn't get redefined and did a little housekeeping

File: src/main.rs [hunk 1]
```diff
@@ -1,80 +1,217 @@
...
     io::{prelude::*, BufReader},
     net::{TcpListener, TcpStream},
 };
 
 type Handler = fn(&str) -> String;
 
-const body_404: &str = include_str!("404.html");
-const body_index: &str = include_str!("index.html");
+const BODY_404: &str = include_str!("404.html");
+const BODY_INDEX: &str = include_str!("index.html");
```
Comment (twny) [inline src/main.rs]: good catch constants should be all caps

File: src/main.rs [hunk 2]
```diff
@@ -1,80 +1,217 @@
...
-fn get_foo(_: &str) -> String {
+fn handler_foo(_: &str) -> String {
     let status = "HTTP/1.1 404 Not Found \r\n";
     let body = "<html><h1>Foo</h1></html>";
     let size = format!("Content-Length: {}\r\n", body.len());
     let response = format!("{status}{size}\r\n{body}");
-    return response;
+
+    response
 }
```
Comment (twny) [inline src/main.rs]: i don't think we need this code anymore

**Comment by trvswgnr** [comment]
- resolved conflicts with most recent changes
- updated to use `Arc` for sharing `HashMap` between threads without creating a new object every time
- changed hashmap to use `&str` instead of String

**Comment by twny** [comment]
this is so legit! i can't wait to review this on stream tomorrow and merge it. then we can add thread pools !

File: src/main.rs [hunk 3]
```diff
@@ -40,44 +42,177 @@ fn main() {
...
     for stream in listener.incoming() {
+        /* Arc clone let's us create a new reference to the same data
+        instead of cloning the data itself */
+        let routes = Arc::clone(&routes);
         let stream = stream.unwrap();
 
         // TODO make this a thead pool
-        thread::spawn(|| {
-            handle_connection(stream);
+        thread::spawn(move || {
```
Comment (twny) [inline src/main.rs]:
```suggestion
        thread::spawn(|| {
```

File: src/main.rs [hunk 4]
```diff
@@ -40,44 +42,177 @@ fn main() {
...
+    let mut parts = request_line.split_whitespace();
+    let method = parts.next().unwrap();
+    let route = parts.next().unwrap();
+    let version = parts.next().unwrap();
+    let mut route_parts = route.split('?');
+    let route = route_parts.next().unwrap();
+    let query = route_parts.next().unwrap_or("");
+    let mut headers = HashMap::new();
+
+    for (index, header) in request.iter().enumerate() {
```
Comment (twny) [inline src/main.rs:123]:
```suggestion
    for (index, header) in request.iter().enumerate().skip(1) {
```

File: src/main.rs [hunk 5]
```diff
@@ -40,44 +42,177 @@ fn main() {
...
+        if index > 0 {
```
Comment (twny) [inline src/main.rs:124]:
```suggestion
```

File: src/main.rs [hunk 6]
```diff
@@ -40,44 +42,177 @@ fn main() {
...
+                continue;
+            }
+            let mut split = header.split(':');
+            headers.insert(
+                // key
+                split.next().unwrap().trim().to_string(),
+                // value
+                split.collect::<Vec<&str>>().join(":").trim().to_string(),
+            );
+        }
```
Comment (twny) [inline src/main.rs:135]:
```suggestion
```

Comment (twny) [inline src/main.rs:135]:
```suggestion
        if !header.contains(':') || header.starts_with('{') {
            continue;
        }
        let mut split = header.split(':');
        headers.insert(
            // key
            split.next().unwrap().trim().to_string(),
            // value
            split.collect::<Vec<&str>>().join(":").trim().to_string(),
        );
    }
```

**Pull request merged**